packages/server-wallet/src/onchain-service/index.ts

```diff
@@ -0,0 +1,228 @@
+  transactionRetries?: number; // Maximum number of times a tx is retried
+};
+
+// This is used instead of the ethers `Transaction` because that type
+// requires the nonce and chain ID to be specified, when sometimes those
+// arguments are not known at the time of creating a transaction.
+type MinimalTransaction = {
+  to: Address;
+  value: string;
+  data: string;
+};
+
+type AssetHolderInformation = {
+  assetId: Address;
+  contractAddress: Address;
+};
+
+export interface OnchainServiceInterface {
+  // NOTE: will return void and throw an error in final implementation
+  registerChannel(
+    channelId: Bytes32,
+    assetHolders: AssetHolderInformation[]
+  ): Promise<string | undefined>;
+  submitTransaction(
+    channelId: Bytes32,
+    tx: MinimalTransaction
+  ): Promise<providers.TransactionResponse>;
+
+  // TODO: remove in v1
+  attachChannelWallet(wallet: ChannelWallet): void;
+}
+
+export class OnchainService implements OnchainServiceInterface {
+  private provider: providers.JsonRpcProvider;
+  private wallet: Wallet;
+  private retries: number;
+
+  // NOTE: V0 has an in-memory store, which will eventually be
+  // shifted to more permanent storage
+  // Holds latest channel event by type (i.e. { "fundsDeposited": {...} })
+  private store: Map<string, object> = new Map();
+
+  // Stores references to all contracts in memory
+  private assetHolders: Map<string, AssetHolderInformation & {contract: Contract}> = new Map();
+
+  // Transaction queue
+  private queue = new PriorityQueue({concurrency: 1});
+
+  // Memory tracking wallet nonce to assist with retries
+  private memoryNonce = 0;
+
+  // TODO: Remove reference to this object, and instead replace with
+  // outbox based communication. Right now used to call correct callbacks
+  private channelWallet: ChannelWallet | undefined = undefined;
+
+  constructor(env: OnchainServiceEnvironment) {
+    this.retries = env.transactionRetries || DEFAULT_MAX_RETRIES;
+    this.provider =
+      typeof env.provider === 'string' ? new providers.JsonRpcProvider(env.provider) : env.provider;
+    this.wallet =
+      typeof env.wallet === 'string'
+        ? new Wallet(env.wallet, this.provider)
+        : (env.wallet as Wallet).connect(this.provider);
+  }
+
+  public attachChannelWallet(wallet: ChannelWallet): void {
+    this.channelWallet = wallet;
+  }
+
+  /**
+   * Adds a channel to watch events and submit transactions for
+   *
+   * @param channelId Unique channel identifier to register chain listeners for
+   * @param assetHolders Onchain addresses of the channel chain context
+   * @returns String if there is an error, undefined if successful
+   */
+  public registerChannel(
+    channelId: Bytes32,
+    assetHolders: AssetHolderInformation[]
+  ): Promise<string | undefined> {
+    if (!this.channelWallet) {
+      throw new Error(
+        'Must attach channel wallet before registering channel, this will go away in v1'
+      );
+    }
```

Comment:
What does it mean to register a channel if you don't have a wallet you can take action on or call a function on once the event is fired? I'm not parroting events from this service in any way, so it would only update storage (which may be desirable, I'm not sure).